### PR TITLE
#7012 Can no longer reactivate or void sales orders with exported shi…

### PR DESCRIPTION
**Pull request by dragospodariu96**
#7012

### CODE REVIEW

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+;
+
+-- 2020-07-21T08:48:58.721Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message SET EntityType='de.metas.inout',Updated=TO_TIMESTAMP('2020-07-21 11:48:58','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Message_ID=544987
+;
+
+-- 2020-07-21T08:53:06.080Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message SET MsgText='Der Beleg kann nicht geändert werden, weil bereit Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:06','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Message_ID=544987
```
Comment (metas-ts) [inline backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql:23]:
```suggestion
UPDATE AD_Message SET MsgText='Der Beleg kann nicht geändert werden, weil bereits Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:06','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Message_ID=544987
```

(fixing typo in the text i gave to you 😊 )

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+;
+
+-- 2020-07-21T08:53:40.911Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message_Trl SET MsgText='Der Beleg kann nicht geändert werden, weil bereit Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:40','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='de_CH' AND AD_Message_ID=544987
+;
+
+-- 2020-07-21T08:53:46.026Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message_Trl SET MsgText='Der Beleg kann nicht geändert werden, weil bereit Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:46','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='de_DE' AND AD_Message_ID=544987
```
Comment (metas-ts) [inline backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql:38]:
```suggestion
UPDATE AD_Message_Trl SET MsgText='Der Beleg kann nicht geändert werden, weil bereits Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:46','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='de_DE' AND AD_Message_ID=544987
```

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+;
+
+-- 2020-07-21T08:53:32.534Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message_Trl SET IsTranslated='Y', MsgText='Changing the document is not allowed because shipment schedules were already exported.',Updated=TO_TIMESTAMP('2020-07-21 11:53:32','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='en_US' AND AD_Message_ID=544987
+;
+
+-- 2020-07-21T08:53:40.911Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message_Trl SET MsgText='Der Beleg kann nicht geändert werden, weil bereit Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:40','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='de_CH' AND AD_Message_ID=544987
```
Comment (metas-ts) [inline backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql:33]:
```suggestion
UPDATE AD_Message_Trl SET MsgText='Der Beleg kann nicht geändert werden, weil bereits Lieferdispo-Datensätze exportiert wurden.',Updated=TO_TIMESTAMP('2020-07-21 11:53:40','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Language='de_CH' AND AD_Message_ID=544987
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/IShipmentSchedulePA.java [hunk 4]
```diff
@@ -53,6 +50,8 @@
 	@Nullable
 	ShipmentScheduleId getShipmentScheduleIdByOrderLineId(OrderLineId orderLineId);
 
+	boolean existsExportedShipmentForOrder(@NonNull final OrderId orderId);
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/IShipmentSchedulePA.java:53]:
```suggestion
	boolean existsExportedShipmentScheduleForOrder(@NonNull final OrderId orderId);
```

Shipment != ShipmentSchedule

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java [hunk 5]
```diff
@@ -148,6 +165,19 @@ public ShipmentScheduleId getShipmentScheduleIdByOrderLineId(@NonNull final Orde
...
 
+	@Override
+	public boolean existsExportedShipmentForOrder(@NonNull final OrderId orderId)
+	{
+		return queryBL
+				.createQueryBuilder(I_M_ShipmentSchedule.class)
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_M_ShipmentSchedule.COLUMNNAME_C_Order_ID, orderId)
+				.addInArrayFilter(I_M_ShipmentSchedule.COLUMNNAME_ExportStatus, UnallowedExportStatuses.values())
+				.orderBy(I_M_ShipmentSchedule.COLUMNNAME_M_ShipmentSchedule_ID)
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java:176]:
```suggestion
```

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+;
+
+-- 2020-07-21T07:54:21.338Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+INSERT INTO AD_Message_Trl (AD_Language,AD_Message_ID, MsgText,MsgTip, IsTranslated,AD_Client_ID,AD_Org_ID,Created,Createdby,Updated,UpdatedBy) SELECT l.AD_Language, t.AD_Message_ID, t.MsgText,t.MsgTip, 'N',t.AD_Client_ID,t.AD_Org_ID,t.Created,t.Createdby,t.Updated,t.UpdatedBy FROM AD_Language l, AD_Message t WHERE l.IsActive='Y'AND (l.IsSystemLanguage='Y' AND l.IsBaseLanguage='N') AND t.AD_Message_ID=544987 AND NOT EXISTS (SELECT 1 FROM AD_Message_Trl tt WHERE tt.AD_Language=l.AD_Language AND tt.AD_Message_ID=t.AD_Message_ID)
+;
+
+-- 2020-07-21T07:54:49.689Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Message SET Value='webui.salesorder.shipmentschedule.exported',Updated=TO_TIMESTAMP('2020-07-21 10:54:49','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Message_ID=544987
```
Comment (metas-ts) [inline backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5563930_sys_gh7012ValidationMessageForSalesOrderNotReactivableOrVoidable.sql:13]: what's the point of the `webui.` prefix?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java [hunk 7]
```diff
@@ -433,6 +445,18 @@ public void updatePreparationDate_Override(final Timestamp preparationDate, fina
 		);
 	}
 
+	@Override
+	public void updateExportStatus(final String exportStatus, final PInstanceId pinstanceId)
+	{
+		updateColumnForSelection(
+				I_M_ShipmentSchedule.COLUMNNAME_ExportStatus,
+				exportStatus,
+				false,
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java:454]:
```suggestion
				false /* updateOnlyIfNull */,
```
most IDEs & Editors don't display the parameter name

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java [hunk 8]
```diff
@@ -433,6 +445,18 @@ public void updatePreparationDate_Override(final Timestamp preparationDate, fina
...
+				pinstanceId,
+				true
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java:456]:
```suggestion
				false /* invalidate */
```

i don't see the need to invalidate..why invalidate=true?
PS..proposing change to false..

Comment (dragospodariu96) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentSchedulePA.java]: initially thought a cache reset is needed, seems it's not!

**Pull request merged**